### PR TITLE
Disable fusing ops like slice, pad, cat, take_along_axis unless inputs are fusion inputs

### PR DESCRIPTION
**Pull request by naoyam**
For now, those ops require inputs to be fusion inputs by default, so they will cause segmentation.

To enable more aggressive fusion, use `NVFUSER_ENABLE=memory_promotion`.

Workarounds #458 

Also related to #284

### CODE REVIEW

**Comment by naoyam** [comment]
!build

File: csrc/device_lower/analysis/predicate_elimination.cpp [hunk 1]
```diff
@@ -91,7 +91,6 @@ class PredicateAnalyzer : public OptOutDispatch {
 
     for (auto id : consumer->getLeafDomain()) {
       if (analyzer.needsPredicate(id)) {
-        producer->fusion()->print();
```
Comment (naoyam) [inline csrc/device_lower/analysis/predicate_elimination.cpp:94]: This was accidentally left.

**Comment by zasdfgbnm** [comment]
Also fixes https://github.com/NVIDIA/Fuser/issues/548

File: csrc/scheduler/registry.cpp [hunk 2]
```diff
@@ -63,17 +63,43 @@ bool rejectScheduleFusionInputRequirement(
...
+      // promotion, i.e., persistent global sync in the case of
+      // block-parallel ops, it needs to be explictly enabled unless
+      // it's for the persistent scheuduler, which is already
+      // persistent. For block-parallel ops, it means the problem size
+      // is small enough to fit in registers, so the required capacity
+      // of the shared memory is likely small enough.
+      if (expr->isA<TorchGatherOp>() &&
+          expr->as<TorchGatherOp>()->exactSizes() &&
+          (isOptionEnabled(EnableOption::MemoryPromotion) ||
+           schedule_strategy == ScheduleHeuristic::Persistent)) {
```
Comment (zasdfgbnm) [inline csrc/scheduler/registry.cpp:83]: I don't know how our persistent kernel is implemented, but does persistent kernel implies cooperative launch? For example, if I have a tensor of shape (100000000, 1024), and I want to do a normalization on the second dim, can I just launch 100000000 blocks and each block do block reduction and broadcasting? For this case, I don't think we have block sync, but if we have `take_along_axis` along the first dim, does the memory promotion create block sync, which then will fail?

Comment (naoyam) [reply csrc/scheduler/registry.cpp:83]: Oh, I'm sorry but that was just I forgot removing a temporary change I was playing with. `schedule_strategy == ScheduleHeuristic::Persistent` should be removed. I'll update the PR and rerun the tests.

Comment (naoyam) [reply csrc/scheduler/registry.cpp]: Updated

File: csrc/scheduler/registry.cpp [hunk 3]
```diff
@@ -63,17 +63,42 @@ bool rejectScheduleFusionInputRequirement(
...
-          !expr->as<TorchGatherOp>()->exactSizes())) &&
-        rejectScheduleFusionInputRequirement(expr, schedule_strategy)) {
-      return true;
+    if (expr->isOneOf<SelectOp, IndexSelectOp, TorchGatherOp>()) {
+      // For now, only relax the input requirement when it's
+      // take_along_axis. Also since this would require memory
+      // promotion, i.e., persistent global sync in the case of
+      // block-parallel ops, it needs to be explictly enabled unless
+      // it's for the persistent scheuduler, which is already
+      // persistent. For block-parallel ops, it means the problem size
```
Comment (zasdfgbnm) [inline csrc/scheduler/registry.cpp:77]: Should the comment be updated as well?

**Comment by jjsjann123** [comment]
nitpick, would it be easy to have a knob to turn on/off this requirement?

I can always just revert the PR locally, but I guess that might gets tricky with future changes stacked on.

File: csrc/scheduler/registry.cpp (hunk 3 above)
Comment (naoyam) [reply csrc/scheduler/registry.cpp]: Yes, thanks, updated.

**Comment by naoyam** [comment]
> nitpick, would it be easy to have a knob to turn on/off this requirement?
> 
> I can always just revert the PR locally, but I guess that might gets tricky with future changes stacked on.

Yes, you can use `NVFUSER_ENABLE=memory_promotion` to enable fusions.

**Review by zasdfgbnm — APPROVED** [review]

**Pull request merged**